File: packages/shared/src/schemas/provider-key.schemas.ts
import { z } from '@hono/zod-openapi'

export const AI_API_PROVIDERS = [
  'openai',
  'openrouter',
  'lmstudio',
  'ollama',
  'xai',
  'google_gemini',
  'anthropic',
  'groq',
  'together'
] as const
export const aiProviderSchema = z.enum(AI_API_PROVIDERS)

export const providerSchema = z.enum(AI_API_PROVIDERS)
export type APIProviders = z.infer<typeof providerSchema>

export const ProviderKeySchema = z
  .object({
    id: z.number().openapi({ example: 1716537600000, description: 'Provider Key ID' }),
    provider: z
      .string()
      .openapi({ example: 'openai', description: 'AI Provider identifier (e.g., openai, anthropic)' }),
    // NOTE: We intentionally DO NOT include the 'key' field in the response schema for security.
    // The full key might be returned on creation/update but shouldn't be listed.
    key: z
      .string()
      .openapi({ example: 'sk-xxxxxxxxxxxxxxxxxxxx', description: 'The actual API Key (handle with care)' }),
    created: z.number().openapi({ example: 1716537600000, description: 'Creation timestamp (ISO 8601)' }),
    updated: z.number().openapi({ example: 1716537600000, description: 'Last update timestamp (ISO 8601)' })
  })
  .openapi('ProviderKey')

export const CreateProviderKeyInputSchema = ProviderKeySchema.omit({
  id: true,
  created: true,
  updated: true
})

export type CreateProviderKeyInput = z.infer<typeof CreateProviderKeyInputSchema>

export const ProviderKeyWithSecretSchema = ProviderKeySchema.extend({
  key: z.string().openapi({ example: 'sk-xxxxxxxxxxxxxxxxxxxx', description: 'The actual API Key (handle with care)' })
}).openapi('ProviderKeyWithSecret')

export const CreateProviderKeyBodySchema = z
  .object({
    provider: z.string().min(1).openapi({ example: 'anthropic' }),
    key: z.string().min(1).openapi({ example: 'sk-ant-xxxxxxxx' })
  })
  .openapi('CreateProviderKeyRequestBody')

export const UpdateProviderKeyBodySchema = z
  .object({
    provider: z.string().min(1).optional().openapi({ example: 'google' }),
    key: z.string().min(1).optional().openapi({ example: 'aizaxxxxxxxxxxxxx' })
  })
  .refine((data) => data.provider || data.key, {
    message: 'At least one of provider or key must be provided for update'
  })
  .openapi('UpdateProviderKeyRequestBody')

// --- Request Parameter Schemas ---
export const ProviderKeyIdParamsSchema = z
  .object({
    keyId: z
      .string()
      .min(1)
      .openapi({
        param: { name: 'keyId', in: 'path' },
        example: 'key-1a2b3c4d',
        description: 'The ID of the provider key'
      })
  })
  .openapi('ProviderKeyIdParams')

export const ProviderKeyResponseSchema = z
  .object({
    success: z.literal(true),
    data: ProviderKeyWithSecretSchema
  })
  .openapi('ProviderKeyResponse')

export const ProviderKeyListResponseSchema = z
  .object({
    success: z.literal(true),
    data: z.array(ProviderKeySchema)
  })
  .openapi('ProviderKeyListResponse')

export type CreateProviderKeyBody = z.infer<typeof CreateProviderKeyBodySchema>
export type UpdateProviderKeyBody = z.infer<typeof UpdateProviderKeyBodySchema>
export type ProviderKeyIdParams = z.infer<typeof ProviderKeyIdParamsSchema>

export type ProviderKey = z.infer<typeof ProviderKeySchema>

export type UpdateProviderKeyInput = z.infer<typeof UpdateProviderKeyBodySchema>

File: packages/python_backend/app/schemas/provider_key_schemas.py
from typing import Optional, List, Dict, Any, Literal
from pydantic import BaseModel, Field, ConfigDict, model_validator
from enum import Enum
from app.utils.storage_timestamp_utils import convert_timestamp_to_ms_int

class AIProviderEnum(str, Enum):
    OPENAI = "openai"
    OPENROUTER = "openrouter"
    LMSTUDIO = "lmstudio"
    OLLAMA = "ollama"
    XAI = "xai"
    GOOGLE_GEMINI = "google_gemini"
    ANTHROPIC = "anthropic"
    GROQ = "groq"
    TOGETHER = "together"

class ProviderKey(BaseModel):
    id: int = Field(..., example=1677657600000, description="Provider Key ID (Unix ms)")
    name: str = Field(..., min_length=1, example="My OpenAI Key", description="User-defined name for the key")
    provider: str = Field(..., min_length=1, example="openai", description="AI Provider identifier")
    key: str = Field(..., min_length=1, example="sk-xxxxxxxxxxxxxxxxxxxx", description="The actual API Key")
    is_default: bool = Field(default=False, alias="isDefault", description="Whether this key is the default for its provider")
    created: int = Field(..., example=1677657600000, description="Creation timestamp (Unix ms)")
    updated: int = Field(..., example=1677657900000, description="Last update timestamp (Unix ms)")
    model_config = ConfigDict(title="ProviderKey", populate_by_name=True)

class CreateProviderKeyBody(BaseModel):
    name: str = Field(..., min_length=1, example="My OpenAI Key")
    provider: str = Field(..., min_length=1, example="anthropic")
    key: str = Field(..., min_length=1, example="sk-ant-xxxxxxxx")
    is_default: Optional[bool] = Field(None, alias="isDefault", example=True)
    model_config = ConfigDict(title="CreateProviderKeyRequestBody", populate_by_name=True)

class UpdateProviderKeyBody(BaseModel):
    name: Optional[str] = Field(None, min_length=1, example="My Updated Key Name")
    provider: Optional[str] = Field(None, min_length=1, example="google")
    key: Optional[str] = Field(None, min_length=1, example="aizaxxxxxxxxxxxxx")
    is_default: Optional[bool] = Field(None, alias="isDefault", example=False)
    
    @model_validator(mode='before')
    @classmethod
    def check_at_least_one_field(cls, values):
        if isinstance(values, dict):
            name = values.get('name')
            provider = values.get('provider')
            key = values.get('key')
            is_default = values.get('isDefault') or values.get('is_default')
            
            if name is None and provider is None and key is None and is_default is None:
                raise ValueError("At least one of name, provider, key, or isDefault must be provided for update")
            
            if name == "" or provider == "" or key == "":
                raise ValueError("Fields cannot be empty strings")
                
        return values

    model_config = ConfigDict(title="UpdateProviderKeyRequestBody", populate_by_name=True)

class ProviderKeyResponse(BaseModel):
    success: Literal[True] = True
    data: ProviderKey
    model_config = ConfigDict(title="ProviderKeyResponse")

class ProviderKeyListResponse(BaseModel):
    success: Literal[True] = True
    
    class ProviderKeyListItem(BaseModel):
        id: int = Field(..., example=1677657600000, description="Provider Key ID (Unix ms)")
        name: str = Field(..., example="My OpenAI Key", description="User-defined name for the key")
        provider: str = Field(..., example="openai", description="AI Provider identifier")
        is_default: bool = Field(..., alias="isDefault", description="Whether this key is the default for its provider")
        created: int = Field(..., example=1677657600000, description="Creation timestamp (Unix ms)")
        updated: int = Field(..., example=1677657900000, description="Last update timestamp (Unix ms)")
        model_config = ConfigDict(title="ProviderKeyListItem", populate_by_name=True)
    
    data: List[ProviderKeyListItem]
    model_config = ConfigDict(title="ProviderKeyListResponse")